File: sql-chatbot-rails/lib/sql_chatbot/grammar/template_compiler.rb
# frozen_string_literal: true

require "sql_chatbot/grammar/registry"
require "sql_chatbot/grammar/primitives"
require "sql_chatbot/grammar/modifiers"

module SqlChatbot
  module Grammar
    module TemplateCompiler
      def self.compile(intent, registry)
        return { ok: false, reason: "unmatched: #{intent[:reason]}" } if intent[:status].to_s == "unmatched"

        entity_name = registry.aliases[intent[:entity]] || intent[:entity]
        entity = registry.entities[entity_name]
        return { ok: false, reason: "entity '#{intent[:entity]}' not in registry" } unless entity

        begin
          modifiers = Array(intent[:modifiers])
          primitive_sym = intent[:primitive].to_s
          rank_field = intent[:rank_field]
          limit_n    = intent[:n]

          # TOP_N has its own ORDER BY + LIMIT baked in. If the intent also
          # includes order_by / limit modifiers, they describe how TOP_N should
          # rank — absorb them into the primitive instead of appending, which
          # would produce duplicate ORDER BY / LIMIT clauses.
          if primitive_sym == "TOP_N"
            order_mod = modifiers.find { |m| (m[:kind] || m["kind"]).to_s == "order_by" }
            limit_mod = modifiers.find { |m| (m[:kind] || m["kind"]).to_s == "limit" }
            if order_mod
              rank_field ||= order_mod[:field] || order_mod["field"]
              modifiers = modifiers.reject { |m| m.equal?(order_mod) }
            end
            if limit_mod
              limit_n ||= limit_mod[:value] || limit_mod["value"]
              modifiers = modifiers.reject { |m| m.equal?(limit_mod) }
            end
          end

          sql = Primitives.build(
            primitive:  intent[:primitive],
            entity:     entity,
            field:      intent[:field],
            which:      intent[:which],
            n:          limit_n,
            rank_field: rank_field,
            group_by:   intent[:group_by]
          )

          modifiers.each do |m|
            sql = Modifiers.apply(sql, m, entity)
          end

          deleted_col = entity.timestamps["deleted"] || entity.timestamps[:deleted]
          if deleted_col
            sql = with_soft_delete(sql, entity, deleted_col)
          end

          unless sql =~ /LIMIT \d+/i || primitive_sym == "COUNT" || sql =~ /COUNT\(/i
            sql = "#{sql} LIMIT 100"
          end

          { ok: true, sql: sql }
        rescue => e
          { ok: false, reason: e.message }
        end
      end

      def self.with_soft_delete(sql, entity, col)
        quoted_ref = Primitives.qc(entity.table, col)
        clause = "#{quoted_ref} IS NULL"

        # Skip if already filtered on this column (match either quoted or unquoted form for safety)
        return sql if sql.include?(quoted_ref)
        return sql if /\b#{Regexp.escape(entity.table)}\.#{Regexp.escape(col)}\b/i.match?(sql)

        if /\bWHERE\b/i.match?(sql)
          return sql.sub(/\bWHERE\b/i) { "WHERE #{clause} AND " }
        end

        before_match = sql.match(/ (GROUP BY|ORDER BY|LIMIT) /i)
        if before_match
          return sql.sub(before_match[0]) { " WHERE #{clause}#{before_match[0]}" }
        end

        "#{sql} WHERE #{clause}"
      end
    end
  end
end

File: packages/agent/src/grammar/template-compiler.ts
import { Registry, Entity } from './registry.js';
import { buildPrimitive, PrimitiveKind, qc } from './primitives.js';
import { applyModifier, Modifier } from './modifiers.js';

export type Intent =
  | {
      status: 'matched';
      primitive: PrimitiveKind;
      entity: string;
      modifiers: Modifier[];
      field?: string;
      which?: 'MIN' | 'MAX';
      n?: number;
      rankField?: string;
      groupBy?: string;
      confidence: number;
    }
  | { status: 'unmatched'; confidence: number; reason: string };

export type CompileResult =
  | { ok: true; sql: string }
  | { ok: false; reason: string };

export function compileTemplate(intent: Intent, registry: Registry): CompileResult {
  if (intent.status === 'unmatched') {
    return { ok: false, reason: `unmatched: ${intent.reason}` };
  }

  const entityName = registry.aliases[intent.entity] ?? intent.entity;
  const entity = registry.entities[entityName];
  if (!entity) return { ok: false, reason: `entity '${intent.entity}' not in registry` };

  try {
    let sql = buildPrimitive({
      primitive: intent.primitive,
      entity,
      field: intent.field,
      which: intent.which,
      n: intent.n,
      rankField: intent.rankField,
      groupBy: intent.groupBy,
    });

    for (const m of intent.modifiers) sql = applyModifier(sql, m, entity);

    if (entity.timestamps.deleted) {
      sql = withSoftDelete(sql, entity);
    }

    if (!/LIMIT \d+/i.test(sql) && intent.primitive !== 'COUNT' && !/COUNT\(/i.test(sql)) {
      sql = `${sql} LIMIT 100`;
    }

    return { ok: true, sql };
  } catch (e) {
    return { ok: false, reason: (e as Error).message };
  }
}

function withSoftDelete(sql: string, entity: Entity): string {
  const col = entity.timestamps.deleted!;
  const quotedRef = qc(entity.table, col);
  const clause = `${quotedRef} IS NULL`;
  // If the column reference already appears (user explicitly filtered on it), skip.
  // Match either the quoted form or the unquoted-pair form (defensive).
  if (sql.includes(quotedRef) || new RegExp(`\\b${entity.table}\\.${col}\\b`, 'i').test(sql)) return sql;
  if (/ WHERE /i.test(sql)) return sql.replace(/ WHERE /i, ` WHERE ${clause} AND `);
  // No WHERE: inject before GROUP BY/ORDER BY/LIMIT, or append.
  const beforeGroupOrOrder = sql.match(/ (GROUP BY|ORDER BY|LIMIT) /i);
  if (beforeGroupOrOrder) return sql.replace(beforeGroupOrOrder[0], ` WHERE ${clause}${beforeGroupOrOrder[0]}`);
  return `${sql} WHERE ${clause}`;
}
